server/api/oauth2.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+
+	"github.com/mattermost/mattermost-server/model"
+
+	"github.com/mattermost/mattermost-plugin-msoffice/server/store"
+)
+
+func (api *api) InitOAuth2(userID string) (url string, err error) {
+	conf := api.Remote.NewOAuth2Config()
+	state := fmt.Sprintf("%v_%v", model.NewId()[0:15], userID)
+	err = api.OAuth2StateStore.StoreOAuth2State(state)
+	if err != nil {
+		return "", err
+	}
+
+	return conf.AuthCodeURL(state, oauth2.AccessTypeOffline), nil
+}
+
+func (api *api) CompleteOAuth2(authedUserID, code, state string) error {
+	oconf := api.Remote.NewOAuth2Config()
+
+	err := api.OAuth2StateStore.VerifyOAuth2State(state)
+	if err != nil {
+		return errors.WithMessage(err, "missing stored state")
+	}
+
+	mattermostUserID := strings.Split(state, "_")[1]
+	if mattermostUserID != authedUserID {
+		return errors.New("not authorized, user ID mismatch")
+	}
+
+	ctx := context.Background()
+	tok, err := oconf.Exchange(ctx, code)
+	if err != nil {
+		return err
+	}
+
+	client := api.Remote.NewClient(ctx, tok)
+	me, err := client.GetMe()
+	if err != nil {
+		return err
+	}
+
+	u := &store.User{
+		PluginVersion:    api.Config.PluginVersion,
+		MattermostUserID: mattermostUserID,
+		Remote:           me,
+		OAuth2Token:      tok,
+	}
+
+	err = api.UserStore.StoreUser(u)
+	if err != nil {
+		return err
+	}
+
+	message := fmt.Sprintf("### Welcome to the Microsoft Office plugin!\n"+
+		"Here is some info to prove we got you logged in\n"+
+		"Name: %s \n", me.DisplayName)
+	api.Poster.PostDirect(mattermostUserID, message, "custom_TODO")
```

Comment:
Just so I understand, what is `custom_TODO`? Or is this a TODO in and of itself?

Reply:
removed the TODO comment for now, recurrence will come back soon.